plugins/aws/providers.go

```diff
@@ -0,0 +1,142 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/1Password/shell-plugins/sdk"
+	"github.com/1Password/shell-plugins/sdk/schema/fieldname"
+	confighelpers "github.com/99designs/aws-vault/v7/vault"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+)
+
+type assumeRoleProvider struct {
+	confighelpers.AssumeRoleProvider
+	stsCacheWriter
+}
+
+func (p assumeRoleProvider) Retrieve(ctx context.Context) (aws.Credentials, error) {
+	credentials, err := p.AssumeRoleProvider.Retrieve(ctx)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	err = p.stsCacheWriter.persist(credentials)
+	if err != nil {
+		return aws.Credentials{}, err
+	}
+
+	return credentials, nil
+}
+
+func NewAssumeRoleProvider(awsConfig *confighelpers.Config, in sdk.ProvisionInput, out *sdk.ProvisionOutput) aws.CredentialsProvider {
+	roleCacheKey := getRoleCacheKey(awsConfig.RoleARN)
```

Comment:
Is this unique enough? What if the user has 2 credentials that can assume the same role? They would lead to different assumed roles, but would share 1 cache entry.

Reply:
Plugin cache entries are specific to a certain item. In the case the user specifies the other pair of credentials in another item, we are covered. However, in the (arguably rare) case where a user would just rotate their access keys in the same item, we could indeed run into cache collisions. We could use the ACCESS_KEY_ID value to further isolate entries in this case. ( Access key ids are not sensitive by themselves, but we shouldn't even care about that because these plugin defined keys are encrypted anyway)

Reply:
Introduced Access Key Id as to further differentiate cache keys both for assume role and mfa

Reply:
Ah, good point about it being linked to the item. I was not actively aware of that. That indeed makes this a lot less problematic.

Technically, the Access Key ID is a bit too over-specific since it can change even if the user does not change, but it should work fine in practice and make it slightly more robust :+1: After all, it's not a problem if we drop all cache entries if someone rotates their access key.